### ISSUE TITLE
Allow more flexible Stored Request storage schemes 

### ISSUE THREAD

**Issue by dbemiller** [prebid/prebid-server#366]
Currently, there are two "types" of stored requests. Some are "request-scoped". For example:

```
{
  ext: {
    prebid: {
      targeting: {
        pricegranularity: "low"
      }
    }
  }
}
```

Others are "imp-scoped". For example:

```
{
  banner: {
    format: [
      { w: 300, h: 250 }
    ]
  },
  ext: {
    appnexus: {
      placementId: 123
    }
  }
}
```

A single HTTP request may include several stored request IDs, and make use of both types at once.

However, the PBS Postgres code today constrains host companies who manage Stored Request data in a few ways:

1. It must be able to fetch all the data in a single SQL query.
2. The query should include an `%ID_LIST%` macro which will be resolved to a list of _all_ the Stored Requests.

I strongly support (1), for performance reasons. However, (2) is less reasonable, and doesn't really have any good reasons motivating it.

In particular, it forces the IDs for "request-scoped" data and "imp-scoped" data to be globally unique. It's not possible (as far as I can tell) to build a SQL query which chooses the right ID in all situations. There isn't much value to this constraint, so it just makes the data unnecessarily difficult to manage.

This ticket is for designing (& possibly building) a feature to relax (2), but only if we can do it _without_ sacrificing the performance benefits of (1).

**Comment by hhhjort** [comment]
Do we know of any benchmarks that would talk to the performance benefit of constraint 1? As I see it, you gain not having a second connection for the second query as a pro, but the con would be Postgres would have to figure out how to parallelize the two subqueries, and then join them, rather than being given the two independent subqueries to run with.

I am not sold that 1 gives us a significant performance benefit. A decent database implementation should be good at efficiently splitting up the compound query, but any decent database client should be running a connection pool which would mitigate the second connection cost. It may be 1 is a significant performance enhancement, I would just like to see some experience or numbers to back this up. I would hate for us to bend over backwards to keep 1, just to find out it saves us only 0.01 ms on each auction.

**Comment by dbemiller** [comment]
No benchmarks, no. Although _we_ happen to care about postgres, bear in mind that the relevant code could be used with any database connection.

I don't think we'll have to bend over backwards to make this happen, though. To throw a simple suggestion out there, we could just make the query resolve two macros: `%REQUEST_ID_LIST%` and `%IMP_ID_LIST%`.

There might be some edge cases, and I obviously haven't written it yet, but... I imagine that would work.

**Comment by hhhjort** [comment]
My points about good database backends and good database clients should generalize across most possible backends. If we don't have to bend over backwards, then I have no issue keeping constraint 1. 

Edge cases for your suggestion above would include one of the two macros being empty. A possible solution would be having 3 queries, one for each possible case (the 4th where both are empty don't need a query. :) )

**Comment by dbemiller** [comment]
I think this is doable. We'd have to change 

```
FetchRequests(ctx context.Context, ids []string) (map[string]json.RawMessage, []error)
```

into:

```
FetchRequests(ctx context.Context, requestID string, impIDs []string) (json.RawMessage, map[string]json.RawMessage, []error)
```

For Postgres (our only serious backend at the moment), the current code expects the query to return a result set with`{id, impData}` columns. To distinguish between "request" and "imp" types, it would need {id, data, type}.

The project today (in `config/stored_requests.go`) suggests using a query like:

```
SELECT id, requestData FROM stored_requests WHERE id in %ID_LIST%
```

An equivalent query for the same schema would be:

```
SELECT id, requestData, 'request' as type
  FROM stored_requests
  WHERE id = %REQUEST_ID%
  UNION ALL
SELECT id, impData, 'imp' as type
  FROM stored_requests
  WHERE id in %IMP_ID_LIST%
```

Hosts who can't guarantee uniqueness could split their stored requests into separate tables, like:

```
SELECT id, requestData, 'request' as type
  FROM stored_requests
  WHERE id = %REQUEST_ID%
  UNION ALL
SELECT id, impData, 'imp' as type
  FROM stored_imps
  WHERE id in %IMP_ID_LIST%
```

Or use a "type" column:

```
SELECT id, requestData, type
  FROM stored_requests
  WHERE (id = %REQUEST_ID% AND type = 'request')
  OR (id in %IMP_ID_LIST% AND type = 'imp')
```

The code itself will just require that the query give back three columns: `{id, data, type}` where `type` is one of `"request"` or `"imp"`

**Comment by dbemiller** [comment]
One small problem: the code can't reliably replace a `%REQUEST_ID%` macro with "null" in a safe way. Ideally we'd want to turn a query like this:

```
SELECT id, requestData, 'request' as type
  FROM stored_requests
  WHERE id = %REQUEST_ID%
  UNION ALL
SELECT id, impData, 'imp' as type
  FROM stored_requests
  WHERE id in %IMP_ID_LIST%
```

into one like this, when `%REQUEST_ID%` isn't used in some request.

```
SELECT id, impData, 'imp' as type
  FROM stored_requests
  WHERE id in %IMP_ID_LIST%
```

Obviously that's impossible to do reliably. Another issue is that today, the `id` can be typed as `int`, `uuid`, `varchar`, etc and the `%ID_LIST% => ($1, $2, $3)` replacement still works. Ideally we'd keep that flexibility. And even if we decided we didn't need it, we would still have to define a sentinel value for the given type, and rely on host companies to avoid that in their DBs. All this sounds pretty crappy.

Luckily, I found a fairly simple workaround. In Postgres, a `WHERE id IN (NULL)` clause works on all types, and returns empty results. Postgres is also smart about it: I ran `explain SELECT id, requestData FROM stored_requests WHERE id in %ID_LIST%;`  to see how this would perform, and got the following results:

```
QUERY PLAN
-------------------------------------------
Result  (cost=0.00..0.00 rows=0 width=16)
  One-Time Filter: false
  (2 rows)
```

which is exactly what I was hoping for :).

So, that's the design I went with. The Fetcher API becomes:

```
FetchRequests(ctx context.Context, requestIDs []string, impIDs []string) (requestData map[string]json.RawMessage, impData map[string]json.RawMessage, errs []error)
```

and the "new" queries should use `WHERE id IN %REQUEST_ID_LIST%` and `WHERE id IN %IMP_ID_LIST%` to get the right behavior.

That said: I still plan to do some more comprehensive testing before finalizing this. All my initial tests today were very promising though.
